src/muse/carbon_budget.py

```diff
@@ -505,68 +497,58 @@ def min_max_bisect(
-        # ub too small -> decrease low
-        # negative exponent (-(threshold - ub)) < 1
-        pow = -(threshold - lb) / abs(denominator)
-        pow = pow if pow < 1 else 1
-        pow = pow if pow > -1 else -1
-        up = low if ub > lb else up
-        low = low * np.exp(pow)
-        low = low if low > 0.0 else 1e-3
+        # Both prices are too high (emissions too low) -> decrease the lower bound
+        exponent = (ub - threshold) / abs(denominator)  # will be negative
+        exponent = max(exponent, -1)  # cap exponent at -1
+        up = low if (ub >= lb) else up  # new upper bound is price with higher emissions
+        low = low * np.exp(exponent)
 
     if ub > threshold and lb > threshold:
-        # lb too big -> increase up
-        # positive exponent (-(threshold - lb)) < 1
-        pow = -2 * (threshold - lb) / abs(denominator)
-        pow = pow if pow < 1 else 1
-        pow = pow if pow > -2 else -2
-        low = up if lb < ub else low
-        up = up * np.exp(pow)
-        up = up if up > 0.0 else 1e-3
+        # Both prices are too low (emissions too high) -> increase the upper bound
+        exponent = (lb - threshold) / abs(denominator)  # will be positive
+        exponent = min(exponent, 1)  # cap exponent at 1
+        low = up if (ub >= lb) else low  # new lower bound is price with lower emissions
+        up = up * np.exp(exponent)
 
     if ub > threshold and lb < threshold:
-        midpoint = round((low + up) / 2.0, 7)
+        # Threshold is between bounds -> perform bisection
+        midpoint = round((low + up) / 2.0, 2)
         m = bisect_loop(market, sectors, equilibrium, commodities, midpoint)
+        emissions_cache[midpoint] = m
         if m < threshold:
-            # midpoint is smaller than lb
-            # lb is a function of up price
+            # Midpoint price is too high -> becomes new upper bound
             up = midpoint
-
         else:
-            # midpoint is smaller than ub
-            # ub is a function of low price
+            # Midpoint price is too low -> becomes new lower bound
             low = midpoint
 
+    # Inverted bounds (i.e. increasing price leads to increasing emissions)
+    # Unlikely case, but included for completeness
     if ub < threshold and lb > threshold:
-        "Inverted bounds"
-        low1 = up
-        up = low
-        low = low1
-        if m < threshold:
-            # midpoint is smaller than lb
-            # lb is a function of up price
+        midpoint = round((low + up) / 2.0, 2)
+        m = bisect_loop(market, sectors, equilibrium, commodities, midpoint)
+        emissions_cache[midpoint] = m
+        if m > threshold:
+            # Midpoint price is too high -> becomes new upper bound
             up = midpoint
-
         else:
-            # midpoint is smaller than ub
-            # ub is a function of low price
+            # Midpoint price is too low -> becomes new lower bound
             low = midpoint
```

Comment:
I'd suggest to separate these 4 cases into separate functions to ease testing and debugging.